tests/ci/ci.py

```diff
@@ -48,6 +52,561 @@
+                records = self.s3.list_prefix(f"{path}{prefix}", S3_BUILDS_BUCKET)
+                records = [record.split("/")[-1] for record in records]
+                print(f"::group::Cache records for {job_type.value} type jobs")
+                print(records)
+                print("::endgroup::")
+                for file in records:
+                    assert file.endswith(self._RECORD_FILE_EXTENSION)
+                    file = file.replace(self._RECORD_FILE_EXTENSION, "")
+                    release_branch = False
+                    assert len(file.split(self._DIV1)) == 2, f"BUG: {file}"
+                    prefix_extended, job_suffix = file.split(self._DIV1)
+                    assert (
+                        len(prefix_extended.split(self._DIV2)) <= 2
+                    ), f"BUG: {file} {prefix_extended}"
+                    if len(prefix_extended.split(self._DIV2)) > 1:
+                        assert (
+                            prefix_extended.split(self._DIV2)[1]
+                            == self._ATTRIBUTE_RELEASE
+                        ), "BUG, no other attributes currently supported"
+                        release_branch = True
+
+                    job_properties = job_suffix.split(self._DIV2)
+                    assert len(job_properties) >= 4, "BUG"
+                    job_name, job_digest, batch, num_batches = (
+                        self._DIV2.join(job_properties[:-3]),
+                        job_properties[-3],
+                        int(job_properties[-2]),
+                        int(job_properties[-1]),
+                    )
+                    assert (
+                        len(job_digest) == JOB_DIGEST_LEN
+                    ), f"Invalid digest: {job_digest}"
+                    if (
+                        job_name not in self.job_digests
+                        or self.job_digests[job_name] != job_digest
+                    ):
+                        # skip records we are not interested in
+                        continue
+                    record = self.Record(
+                        record_type,
+                        job_name,
+                        job_digest,
+                        batch,
+                        num_batches,
+                        release_branch,
+                        file="",
+                    )
+                    if record.to_str_key() not in cache_list:
+                        cache_list[record.to_str_key()] = record
+                        self.cache_data_fetched = False
+                    elif (
+                        not cache_list[record.to_str_key()].release_branch
+                        and record.release_branch
+                    ):
+                        # replace a non-release record with a release one
+                        cache_list[record.to_str_key()] = record
+                        self.cache_data_fetched = False
+
+        self.cache_updated = True
+        return self
```

Comment:
It looks like we shouldn't return `self` here.

Reply:
it helps writing code in functional style